src/user/tag.ts

```diff
@@ -0,0 +1,86 @@
+import { log } from "../config"
+import { Accessory } from "../accessory"
+import { instanceToClass } from "clone-class"
+import { TagPayload } from "wechaty-puppet/dist/src/schemas/tag"
+
+export class Tag  extends Accessory {
+
+  /**
+   * @private
+   */
+  constructor (
+    public readonly id: string,
+  ) {
+    super()
+    log.silly('Tag', `constructor(${id})`)
+
+    // tslint:disable-next-line:variable-name
+    const MyClass = instanceToClass(this, Tag)
+
+    if (MyClass === Tag) {
+      throw new Error(
+        'Tag class can not be instanciated directly!'
+        + 'See: https://github.com/Chatie/wechaty/issues/1217',
+      )
+    }
+
+    if (!this.puppet) {
+      throw new Error('Tag class can not be instanciated without a puppet!')
+    }
+  }
+
+  public static async findAll<T extends typeof Tag> (
+    this  : T,
+  ): Promise<TagPayload []> {
+    log.verbose('Tag', 'findAll()')
+
+    try {
+      const tagList: TagPayload[] = await this.puppet.allTags()
+
+      return tagList
+    } catch (e) {
+      log.error('Tag', 'this.puppet.contactFindAll() rejected: %s', e.message)
+      return [] // fail safe
+    }
+  }
+
+  public static async find<T extends typeof Tag> (
+    this  : T,
+    query : string
+  ): Promise<TagPayload> {
+    log.verbose('Tag', 'find(%s)', JSON.stringify(query))
+
+    const tag = await this.puppet.searchTag(query)
+
+    return tag
+  }
+
+  /**
+   * show all labels across the login account
+   */
+  public async tagList (): Promise<TagPayload[]> {
```

Comment:
What's the difference between the `tagList` and the `findAll`?

As my understanding, they will all return all the tags, so I'd like to suggest to remove the `tagList` method because it does not contribute any new value for us.

Reply:
Agree!

> findAll()  => find all tags from cache
> tagList() => find all tags from API search

You are right, we do not need to know the data source in Wechaty.